=== modified file 'Makefile'
--- Makefile	2011-09-07 00:12:22 +0000
+++ Makefile	2012-05-07 21:22:58 +0000
@@ -29,6 +29,8 @@
 # Just for this Makefile. Sub-makes will run in parallel if requested.
 .NOTPARALLEL:
 
+DIR := $(dir $(lastword $(MAKEFILE_LIST)))
+
 CSCOPE = cscope
 CHECK = tools/check.sh
 CONFIG = tools/config.py
@@ -69,20 +71,20 @@
 # Autotool (detects compiler features)
 
 autotool $(COMMON_MAKEFILE) $(COMMON_HEADER): $(CONFIG_MAKEFILE)
-	$(AUTOTOOL)
+	$(DIR)$(AUTOTOOL)
 	-[ -f $(COMMON_HEADER_PREV) ] && diff -q $(COMMON_HEADER_PREV) $(COMMON_HEADER) && mv -f $(COMMON_HEADER_PREV) $(COMMON_HEADER)
 
 # Build-time configuration
 
 config_default $(CONFIG_MAKEFILE) $(CONFIG_HEADER): $(CONFIG_RULES)
 ifeq ($(HANDS_OFF),y)
-	$(CONFIG) $< hands-off $(PROFILE)
+	$(DIR)$(CONFIG) $< hands-off $(PROFILE)
 else
-	$(CONFIG) $< default $(PROFILE)
+	$(DIR)$(CONFIG) $< default $(PROFILE)
 endif
 
 config: $(CONFIG_RULES)
-	$(CONFIG) $<
+	$(DIR)$(CONFIG) $<
 
 # Release files
 

=== modified file 'kernel/Makefile'
--- kernel/Makefile	2012-03-30 17:39:25 +0000
+++ kernel/Makefile	2012-05-07 21:09:47 +0000
@@ -29,7 +29,9 @@
 ## Configuration
 #
 
-ROOT_PATH = ..
+DIR := $(dir $(lastword $(MAKEFILE_LIST)))
+
+ROOT_PATH = $(DIR)..
 
 VERSION_DEF = $(ROOT_PATH)/version
 
@@ -47,7 +49,7 @@
 ## Common names
 #
 
-DEPEND = Makefile.depend
+DEPEND = $(ROOT_PATH)/Makefile.depend
 DEPEND_PREV = $(DEPEND).prev
 RAW = kernel.raw
 BIN = kernel.bin
@@ -56,16 +58,16 @@
 MAP_PREV = $(MAP).prev
 DISASM = kernel.disasm
 DUMP = kernel.dump
-REAL_MAP = generic/src/debug/real_map
-
-ABI_INCLUDE = generic/include/abi
-ARCH_INCLUDE = generic/include/arch
-GENARCH_INCLUDE = generic/include/genarch
-
-GENMAP = tools/genmap.py
+REAL_MAP = $(DIR)/generic/src/debug/real_map
+
+ABI_INCLUDE = $(DIR)/generic/include/abi
+ARCH_INCLUDE = $(DIR)/generic/include/arch
+GENARCH_INCLUDE = $(DIR)/generic/include/genarch
+
+GENMAP = $(DIR)/tools/genmap.py
 JOBFILE = $(ROOT_PATH)/tools/jobfile.py
 
-LINK = arch/$(KARCH)/_link.ld
+LINK = $(DIR)/arch/$(KARCH)/_link.ld
 EMPTY_MAP = generic/src/debug/empty_map.o
 SIZEOK_MAP = generic/src/debug/sizeok_map.o
 
@@ -140,8 +142,8 @@
 # Mind the mutual ordering with the initialization of AFLAGS and LFLAGS.
 # The arch Makefile.inc must be included after the initialization.
 #
--include arch/$(KARCH)/Makefile.inc
--include genarch/Makefile.inc
+-include $(DIR)/arch/$(KARCH)/Makefile.inc
+-include $(DIR)/genarch/Makefile.inc
 -include $(DEPEND)
 
 ## The at-sign
@@ -439,13 +441,13 @@
 	-[ -f $(DEPEND_PREV) ] && diff -q $(DEPEND_PREV) $@ && mv -f $(DEPEND_PREV) $@
 
 $(ABI_INCLUDE): ../abi/include/
-	ln -sfn ../../$< $@
+	ln -sfn $< $@
 
 $(ARCH_INCLUDE): arch/$(KARCH)/include/
-	ln -sfn ../../$< $@
+	ln -sfn $< $@
 
 $(GENARCH_INCLUDE): genarch/include/
-	ln -sfn ../../$< $@
+	ln -sfn $< $@
 
 $(COMMON_HEADER_ARCH): $(COMMON_HEADER)
-	ln -sfn ../../../$< $@
+	ln -sfn $< $(DIR)/$@

=== modified file 'tools/config.py'
--- tools/config.py	2011-12-02 17:29:43 +0000
+++ tools/config.py	2012-05-07 21:22:58 +0000
@@ -43,7 +43,8 @@
 RULES_FILE = sys.argv[1]
 MAKEFILE = 'Makefile.config'
 MACROS = 'config.h'
-PRESETS_DIR = 'defaults'
+#PRESETS_DIR = 'defaults'
+PRESETS_DIR = os.path.dirname(__file__)+'/../defaults'
 
 def read_config(fname, config):
 	"Read saved values from last configuration run or a preset file"

