=== modified file 'HelenOS.config'
--- HelenOS.config	2012-04-30 14:51:54 +0000
+++ HelenOS.config	2012-05-29 13:30:48 +0000
@@ -371,6 +371,10 @@
 % Kernel console support
 ! CONFIG_KCONSOLE (y/n)
 
+% Copy kernel console to dummy serial at 0x3F8
+! [PLATFORM=ia32|PLATFORM=amd64] CONFIG_KCONSOLE_DSRLNOUT (n/y)
+! [CONFIG_KCONSOLE_DSRLNOUT=y] CONFIG_DSRLNOUT (y)
+
 % Kernel symbol information
 ! CONFIG_SYMTAB (y/n)
 

=== modified file 'kernel/generic/src/main/main.c'
--- kernel/generic/src/main/main.c	2012-03-02 14:56:14 +0000
+++ kernel/generic/src/main/main.c	2012-05-29 13:37:19 +0000
@@ -88,6 +88,17 @@
 #include <sysinfo/stats.h>
 #include <lib/ra.h>
 
+#ifdef CONFIG_KCONSOLE_DSRLNOUT
+#include <genarch/drivers/dsrln/dsrlnout.h>
+#define COPY_KCONSOLE_TO_DSRLN \
+	do { \
+		outdev_t * dsrlndev = dsrlnout_init((ioport8_t *) 0x3f8); \
+		if (dsrlndev) { \
+			stdout_wire(dsrlndev); \
+		} \
+	} while (false)
+#endif
+
 /** Global configuration structure. */
 config_t config = {
 	.identity_configured = false,
@@ -230,6 +241,9 @@
 	ddi_init();
 	arch_post_mm_init();
 	reserve_init();
+#ifdef CONFIG_KCONSOLE_DSRLNOUT
+	COPY_KCONSOLE_TO_DSRLN;
+#endif
 	arch_pre_smp_init();
 	smp_init();
 	

